File: crates/dewasm-backend-java/units/table/tail_ref.java
// requires: rt/trap
// `call`'s checks, raised at the same point in execution order, but handing back the whole slot so the caller can see whether it owns the tail entry.
Rt.Funcref tailSlot(int index, String ty) {
    if (index < 0 || index >= slots.length) {
        Rt.trap("undefined element");
    }
    Rt.Funcref f = slots[index];
    if (f == null) {
        Rt.trap("uninitialized element");
    }
    if (!f.ty.equals(ty)) {
        Rt.trap("indirect call type mismatch");
    }
    return f;
}

File: crates/dewasm-backend-go/units/table/tail_ref.go
// requires: rt/trap
// `call`'s checks, raised at the same point in execution order, but resolving a tail call rather than an ordinary one.
// A tail entry is bound to the instance that built it and reads *that* instance's parked slots, so it is only returned to its owner; every other caller gets the ordinary func value and completes the call itself, which costs one frame per instance switch.
func (t *Table) tailRef(i uint32, typeKey string, owner any) (any, any) {
    if i >= uint32(len(t.slots)) {
        Rt.trap("undefined element")
    }
    slot := t.slots[i]
    if slot == nil {
        Rt.trap("uninitialized element")
    }
    if slot.ty != typeKey {
        Rt.trap("indirect call type mismatch")
    }
    if slot.body != nil && slot.owner == owner {
        return slot.body, nil
    }
    return nil, slot.fn
}
